src/librustc_mir/borrow_check/move_errors.rs

```diff
@@ -0,0 +1,364 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::hir;
+use rustc::mir::*;
+use rustc::ty::{self, TyCtxt};
+use rustc_errors::DiagnosticBuilder;
+use syntax_pos::Span;
+
+use dataflow::move_paths::{IllegalMoveOrigin, IllegalMoveOriginKind, MoveData};
+use dataflow::move_paths::{LookupResult, MoveError, MovePathIndex};
+use util::borrowck_errors::{BorrowckErrors, Origin};
+
+pub(crate) fn report_move_errors<'gcx, 'tcx>(
+    mir: &Mir<'tcx>,
+    tcx: TyCtxt<'_, 'gcx, 'tcx>,
+    move_errors: Vec<MoveError<'tcx>>,
+    move_data: &MoveData<'tcx>,
+) {
+    MoveErrorCtxt {
+        mir,
+        tcx,
+        move_data,
+    }.report_errors(move_errors);
+}
+
+#[derive(Copy, Clone)]
+struct MoveErrorCtxt<'a, 'gcx: 'tcx, 'tcx: 'a> {
+    mir: &'a Mir<'tcx>,
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    move_data: &'a MoveData<'tcx>,
+}
+
+#[derive(Debug)]
+enum GroupedMoveError<'tcx> {
```

Comment:
I feel like a meta comment on the enum explaining (in general) what we group by and why would be super useful. e.g.

---

often when desugaring a pattern match we may have many individual moves in MIR
that are all part of one operation from the user's point-of-view. For example:

```
let (x, y) = foo()
```

would move `x` from the `0` field of some temporary, and `y` from the `1` field. We group such errors together for cleaner error reporting.